### PR TITLE
[FEATURE] Allow to read precomputed minimizer files in chopper count.

### PR DESCRIPTION
**Pull request by smehringer**
Now wen can compute hll sketches from precomputed binary files. 

hm.. seems like the `add_list_item(..., ...::advanced)` call does not work. I neither appears on the regular help page nor the advanced help page. Seems like a bug in the argument parser.

### CODE REVIEW

**Review by eseiler — CHANGES_REQUESTED** [review]

File: include/chopper/count/count_kmers.hpp [hunk 1]
```diff
@@ -54,9 +54,23 @@ inline void count_kmers(robin_hood::unordered_map<std::string, std::vector<std::
...
                     sketch.add(reinterpret_cast<char*>(&hash), sizeof(hash));
+            }
+            else
+            {
+                for (auto && [seq] : sequence_file_type{filename})
+                    for (auto && hash : seq | seqan3::views::kmer_hash(seqan3::ungapped{config.k}))
+                        sketch.add(reinterpret_cast<char*>(&hash), sizeof(hash));
+            }
+
+        }
```
[the remark this answers is not in the document]
Comment (smehringer) [reply include/chopper/count/count_kmers.hpp:73]: hm but why? I don't think it is performance critical and why prohibit wither or if both obviously works?

[the remark this answers is not in the document]
Comment (eseiler) [reply include/chopper/count/count_kmers.hpp]: In raptor, I enforce it. I don't remember if there was a specific reason.

But if it's mixed, one would use `k-mers` for files, and `minimizer` for preprocessed files.
Wouldn't this affect the sketches?

**Review by eseiler — APPROVED** [review]
See comment.
The comment could also be a follow-up. Or not :)

You can merge after deciding

File: include/chopper/count/count_kmers.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (smehringer) [reply include/chopper/count/count_kmers.hpp]: True I did not think about this difference

**Comment by smehringer** [comment]
Since I added wuite a lot it would be nice if you take another look

File: include/chopper/count/check_filenames.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#include <cassert>
+
+#include <robin_hood.h>
+
+#include <chopper/count/configuration.hpp>
+
+namespace chopper::count
+{
+
+//!\brief Checks the `filename_clusters` for consistent files, wither precomputed or sequence files.
```
Comment (eseiler) [inline include/chopper/count/check_filenames.hpp:12]:
```suggestion
//!\brief Checks the `filename_clusters` for consistent files, either precomputed or sequence files.
```

File: include/chopper/count/check_filenames.hpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                                            " does not have the correct file extension (.minimizer)."
+                                            " Mixing non-/precomputed files is not allowed."};
+            }
+            else if (!config.precomputed_files && path.extension() == ".minimizer")
+            {
+                throw std::invalid_argument{"You are providing sequence files but the file " + path.string() +
+                                            " was identified as a precomputed file (.minimizer)."
+                                            " Mixing non-/precomputed files is not allowed."};
+            }
+        }
```
Comment (eseiler) [inline include/chopper/count/check_filenames.hpp:39]: cpp20 would have [ends_with](https://en.cppreference.com/w/cpp/string/basic_string/ends_with) :D

We could use the implementation in some seqan3 validator:

```
    bool case_insensitive_string_ends_with(std::string_view str, std::string_view suffix) const
    {
        size_t const suffix_length{suffix.size()};
        size_t const str_length{str.size()};
        return suffix_length > str_length ?
               false :
               std::ranges::equal(str.substr(str_length - suffix_length), suffix, [] (char const chr1, char const chr2)
               {
                   return std::tolower(chr1) == std::tolower(chr2);
               });
    }
```

Would have the advantage that it's case-insensitive.

The conditions could be expressed by an xor `config.precomputed_files ^ path.extension() == ".minimizer"`. Then the check which exception is thrown could be inside. But not sure if this makes it faster :)

File: include/chopper/count/count_kmers.hpp [hunk 4]
```diff
@@ -29,6 +29,35 @@ using sequence_file_type = seqan3::sequence_file_input<mytraits,
                                                        seqan3::fields<seqan3::field::seq>,
                                                        seqan3::type_list<seqan3::format_fasta, seqan3::format_fastq>>;
 
+inline void process_sequence_files(std::vector<std::string> const & filenames,
+                            configuration const & config,
+                            sketch::hyperloglog & sketch)
```
Comment (eseiler) [inline include/chopper/count/count_kmers.hpp:34]:
```suggestion
inline void process_sequence_files(std::vector<std::string> const & filenames,
                                   configuration const & config,
                                   sketch::hyperloglog & sketch)
```

File: include/chopper/count/count_kmers.hpp [hunk 5]
```diff
@@ -29,6 +29,35 @@ using sequence_file_type = seqan3::sequence_file_input<mytraits,
...
+{
+    for (auto const & filename : filenames)
+    {
+        for (auto && [seq] : sequence_file_type{filename})
+            for (auto && k_hash : seq | seqan3::views::kmer_hash(seqan3::ungapped{config.k}))
+                sketch.add(reinterpret_cast<char*>(&k_hash), sizeof(k_hash));
+    }
```
Comment (eseiler) [inline include/chopper/count/count_kmers.hpp:41]:
```suggestion
        for (auto && [seq] : sequence_file_type{filename})
            for (auto && k_hash : seq | seqan3::views::kmer_hash(seqan3::ungapped{config.k}))
                sketch.add(reinterpret_cast<char*>(&k_hash), sizeof(k_hash));
```

**Review by eseiler — CHANGES_REQUESTED** [review]

File: include/chopper/count/check_filenames.hpp (hunk 3 above)
Comment (smehringer) [reply include/chopper/count/check_filenames.hpp:39]: I added `case_insensitive_string_ends_with` but I think the performance of checking the filenames is negligible and the two if branches are more expressive than the xor.

**Pull request merged**